### ISSUE TITLE
Rebuild django image with specific CPython version

### ISSUE THREAD

**Issue by hwakabh** [hwakabh/bennu-official#38]
# AsIs
As noted in https://github.com/hwakabh/bennu-official/pull/30#issuecomment-1860739029, CPython CNB by paketo-buildpacks would use 3.10.13 [by default](https://github.com/paketo-buildpacks/cpython/blob/159ded75f1f4ca907c3083ca063c9eabb074a6dc/buildpack.toml#L16), and this does not match app's expected version `3.11.5`, which is specified in `.python-version`.
Meaning, paketo-buildpacks would not detect `.python-version` automatically.

```bash
Paketo Buildpack for CPython 1.9.0
  Resolving CPython version
    Candidate version sources (in priority order):
                -> ""
      <unknown> -> ""

    Selected CPython version (using ): 3.10.13

  Executing build process
    Installing CPython 3.10.13
      Completed in 7.861s
``` 

# ToDo
Rebuild image with the flag `BP_CPYTHON_VERSION` to use specific Python version to build, the image in GHCR need to be updated.
Ref: https://github.com/paketo-buildpacks/cpython#bp_cpython_version

# (Optional) Acceptance criteria

**Comment by hwakabh** [comment]
Evidences of rebuilding with the version of `.python-version` file, and this issue would not be related to any PR, so closed the issue after confirm actual version in build artifacts.

```bash
# Build with flags
% pack build ghcr.io/hwakabh/bennu-official/django:latest --builder paketobuildpacks/builder-jammy-full --env BP_CPYTHON_VERSION=$(cat .python-version) --publish
# (...)
===> ANALYZING
Restoring data for SBOM from previous image
===> DETECTING
6 of 9 buildpacks participating
paketo-buildpacks/ca-certificates 3.6.3
paketo-buildpacks/cpython         1.9.0
paketo-buildpacks/pip             0.18.0
paketo-buildpacks/pip-install     0.6.0
paketo-buildpacks/python-start    0.14.14
paketo-buildpacks/procfile        5.6.4
===> RESTORING
Restoring metadata for "paketo-buildpacks/ca-certificates:helper" from app image
Restoring metadata for "paketo-buildpacks/cpython:cpython" from app image
Restoring metadata for "paketo-buildpacks/pip:pip" from cache
Restoring metadata for "paketo-buildpacks/pip:pip-source" from cache
Restoring metadata for "paketo-buildpacks/pip-install:packages" from app image
Restoring metadata for "paketo-buildpacks/pip-install:cache" from cache
Restoring data for "paketo-buildpacks/cpython:cpython" from cache
Restoring data for "paketo-buildpacks/pip:pip" from cache
Restoring data for "paketo-buildpacks/pip:pip-source" from cache
Restoring data for "paketo-buildpacks/pip-install:cache" from cache
Restoring data for "paketo-buildpacks/pip-install:packages" from cache
Restoring data for SBOM from cache
===> BUILDING

Paketo Buildpack for CA Certificates 3.6.3
  https://github.com/paketo-buildpacks/ca-certificates
  Launch Helper: Reusing cached layer
Paketo Buildpack for CPython 1.9.0
  Resolving CPython version
    Candidate version sources (in priority order):
      BP_CPYTHON_VERSION -> "3.11.5"
                         -> ""
      <unknown>          -> ""

    Selected CPython version (using BP_CPYTHON_VERSION): 3.11.5

  Executing build process
    Installing CPython 3.11.5
      Completed in 7.323s

  Generating SBOM for /layers/paketo-buildpacks_cpython/cpython
      Completed in 1ms


  Configuring build environment
    PYTHONPATH          -> "/layers/paketo-buildpacks_cpython/cpython"
    PYTHONPYCACHEPREFIX -> "/tmp"

  Configuring launch environment
    PYTHONPATH -> "/layers/paketo-buildpacks_cpython/cpython"

Paketo Buildpack for Pip 0.18.0
  Resolving Pip version
    Candidate version sources (in priority order):
      <unknown> -> ""

    Selected Pip version (using <unknown>): 23.2.1

  Reusing cached layer /layers/paketo-buildpacks_pip/pip
  Reusing cached layer /layers/paketo-buildpacks_pip/pip-source
Paketo Buildpack for Pip Install 0.6.0
  Executing build process
    Running 'pip install --exists-action=w --cache-dir=/layers/paketo-buildpacks_pip-install/cache --compile --user --disable-pip-version-check --requirement=requirements.txt'
      Looking in links: /layers/paketo-buildpacks_pip/pip-source
      Collecting asgiref==3.7.2 (from -r requirements.txt (line 1))
        Obtaining dependency information for asgiref==3.7.2 from https://files.pythonhosted.org/packages/9b/80/b9051a4a07ad231558fcd8ffc89232711b4e618c15cb7a392a17384bbeef/asgiref-3.7.2-py3-none-any.whl.metadata
        Using cached asgiref-3.7.2-py3-none-any.whl.metadata (9.2 kB)
      Collecting Django==5.0 (from -r requirements.txt (line 2))
        Obtaining dependency information for Django==5.0 from https://files.pythonhosted.org/packages/ba/c7/61b02c0ef9e129080a8c2bffefb3cb2b9ddddece4c44dc473c1c4f0647c1/Django-5.0-py3-none-any.whl.metadata
        Using cached Django-5.0-py3-none-any.whl.metadata (4.1 kB)
      Collecting gunicorn==21.2.0 (from -r requirements.txt (line 3))
        Obtaining dependency information for gunicorn==21.2.0 from https://files.pythonhosted.org/packages/0e/2a/c3a878eccb100ccddf45c50b6b8db8cf3301a6adede6e31d48e8531cab13/gunicorn-21.2.0-py3-none-any.whl.metadata
        Using cached gunicorn-21.2.0-py3-none-any.whl.metadata (4.1 kB)
      Collecting packaging==23.2 (from -r requirements.txt (line 4))
        Obtaining dependency information for packaging==23.2 from https://files.pythonhosted.org/packages/ec/1a/610693ac4ee14fcdf2d9bf3c493370e4f2ef7ae2e19217d7a237ff42367d/packaging-23.2-py3-none-any.whl.metadata
        Using cached packaging-23.2-py3-none-any.whl.metadata (3.2 kB)
      Collecting PyMySQL==1.1.0 (from -r requirements.txt (line 5))
        Obtaining dependency information for PyMySQL==1.1.0 from https://files.pythonhosted.org/packages/e5/30/20467e39523d0cfc2b6227902d3687a16364307260c75e6a1cb4422b0c62/PyMySQL-1.1.0-py3-none-any.whl.metadata
        Downloading PyMySQL-1.1.0-py3-none-any.whl.metadata (4.4 kB)
      Collecting pytz==2023.3.post1 (from -r requirements.txt (line 6))
        Obtaining dependency information for pytz==2023.3.post1 from https://files.pythonhosted.org/packages/32/4d/aaf7eff5deb402fd9a24a1449a8119f00d74ae9c2efa79f8ef9994261fc2/pytz-2023.3.post1-py2.py3-none-any.whl.metadata
        Using cached pytz-2023.3.post1-py2.py3-none-any.whl.metadata (22 kB)
      Collecting sqlparse==0.4.4 (from -r requirements.txt (line 7))
        Using cached sqlparse-0.4.4-py3-none-any.whl (41 kB)
      Collecting typing_extensions==4.9.0 (from -r requirements.txt (line 8))
        Obtaining dependency information for typing_extensions==4.9.0 from https://files.pythonhosted.org/packages/b7/f4/6a90020cd2d93349b442bfcb657d0dc91eee65491600b2cb1d388bc98e6b/typing_extensions-4.9.0-py3-none-any.whl.metadata
        Using cached typing_extensions-4.9.0-py3-none-any.whl.metadata (3.0 kB)
      Collecting whitenoise==6.6.0 (from -r requirements.txt (line 9))
        Obtaining dependency information for whitenoise==6.6.0 from https://files.pythonhosted.org/packages/67/16/bb488ac8230f1bce94943b6654f2aad566d18aae575c8b6d8a99c78c489e/whitenoise-6.6.0-py3-none-any.whl.metadata
        Using cached whitenoise-6.6.0-py3-none-any.whl.metadata (3.7 kB)
      Using cached asgiref-3.7.2-py3-none-any.whl (24 kB)
      Using cached Django-5.0-py3-none-any.whl (8.1 MB)
      Using cached gunicorn-21.2.0-py3-none-any.whl (80 kB)
      Using cached packaging-23.2-py3-none-any.whl (53 kB)
      Downloading PyMySQL-1.1.0-py3-none-any.whl (44 kB)
         ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.8/44.8 kB 97.3 kB/s eta 0:00:00
      Using cached pytz-2023.3.post1-py2.py3-none-any.whl (502 kB)
      Using cached typing_extensions-4.9.0-py3-none-any.whl (32 kB)
      Using cached whitenoise-6.6.0-py3-none-any.whl (19 kB)
      Installing collected packages: pytz, whitenoise, typing_extensions, sqlparse, PyMySQL, packaging, asgiref, gunicorn, Django
        WARNING: The script sqlformat is installed in '/layers/paketo-buildpacks_pip-install/packages/bin' which is not on PATH.
        Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
        WARNING: The script gunicorn is installed in '/layers/paketo-buildpacks_pip-install/packages/bin' which is not on PATH.
        Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
        WARNING: The script django-admin is installed in '/layers/paketo-buildpacks_pip-install/packages/bin' which is not on PATH.
        Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
      Successfully installed Django-5.0 PyMySQL-1.1.0 asgiref-3.7.2 gunicorn-21.2.0 packaging-23.2 pytz-2023.3.post1 sqlparse-0.4.4 typing_extensions-4.9.0 whitenoise-6.6.0
      Completed in 7.479s

  Generating SBOM for /layers/paketo-buildpacks_pip-install/packages
      Completed in 426ms

  Configuring build environment
    PYTHONPATH -> "/layers/paketo-buildpacks_pip-install/packages/lib/python3.11/site-packages:$PYTHONPATH"

  Configuring launch environment
    PYTHONPATH -> "/layers/paketo-buildpacks_pip-install/packages/lib/python3.11/site-packages:$PYTHONPATH"

Paketo Buildpack for Python Start 0.14.14
  Assigning launch processes:
    web (default): python


Paketo Buildpack for Procfile 5.6.4
  https://github.com/paketo-buildpacks/procfile
  Process types:
    web: gunicorn --bind 0.0.0.0:8000 bennu_official.wsgi
===> EXPORTING
Reusing layers from image 'ghcr.io/hwakabh/bennu-official/django@sha256:9614bc3e441f2e7ac81d4bf4aa8cc7fa17276223462df3240bd93890b292be89'
Reusing layer 'paketo-buildpacks/ca-certificates:helper'
Adding layer 'paketo-buildpacks/cpython:cpython'
Adding layer 'paketo-buildpacks/pip-install:packages'
Adding layer 'buildpacksio/lifecycle:launch.sbom'
Adding 1/1 app layer(s)
Adding layer 'buildpacksio/lifecycle:launcher'
Adding layer 'buildpacksio/lifecycle:config'
Reusing layer 'buildpacksio/lifecycle:process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
Saving ghcr.io/hwakabh/bennu-official/django:latest...
*** Images (sha256:eafc7c3cf340513e6275bf702e44e12675f8c424121208008c6868186f432ca5):
      ghcr.io/hwakabh/bennu-official/django:latest
Adding cache layer 'paketo-buildpacks/cpython:cpython'
Reusing cache layer 'paketo-buildpacks/pip:pip'
Reusing cache layer 'paketo-buildpacks/pip:pip-source'
Adding cache layer 'paketo-buildpacks/pip-install:cache'
Adding cache layer 'paketo-buildpacks/pip-install:packages'
Adding cache layer 'buildpacksio/lifecycle:cache.sbom'
Successfully built image ghcr.io/hwakabh/bennu-official/django:latest
```

**Comment by hwakabh** [comment]
Though `BP_CPYTHON_VERSION` was provided, the actual Python version in the artifacts would be still `3.10.12`, which is even not match with `default-version` of CPython CNB v1.9.0, so need to check further.
Available latest of `3.10.x` is `3.10.13` in buildpack.toml of CPython CNB v1.9.0.
Ref: https://github.com/paketo-buildpacks/cpython/blob/v1.9.0/buildpack.toml#L291-L328

```bash
cnb@6d6e72e205b2:/workspace$ ls -al /layers/paketo-buildpacks_cpython/cpython/bin
total 48
drwxr-xr-x  2 1001 cnb  4096 Jan  1  1980 .
drwxr-xr-x 10 1001 cnb  4096 Jan  1  1980 ..
lrwxrwxrwx  1 1001 cnb     9 Jan  1  1980 2to3 -> 2to3-3.11
-rwxr-xr-x  1 1001 cnb   111 Jan  1  1980 2to3-3.11
lrwxrwxrwx  1 1001 cnb     8 Jan  1  1980 idle3 -> idle3.11
-rwxr-xr-x  1 1001 cnb   109 Jan  1  1980 idle3.11
-rwxr-xr-x  1 1001 cnb   239 Jan  1  1980 pip3
-rwxr-xr-x  1 1001 cnb   239 Jan  1  1980 pip3.11
lrwxrwxrwx  1 1001 cnb     9 Jan  1  1980 pydoc3 -> pydoc3.11
-rwxr-xr-x  1 1001 cnb    94 Jan  1  1980 pydoc3.11
lrwxrwxrwx  1 1001 cnb    53 Jan  1  1980 python -> /layers/paketo-buildpacks_cpython/cpython/bin/python3
lrwxrwxrwx  1 1001 cnb    10 Jan  1  1980 python3 -> python3.11
lrwxrwxrwx  1 1001 cnb    17 Jan  1  1980 python3-config -> python3.11-config
-rwxr-xr-x  1 1001 cnb 14472 Jan  1  1980 python3.11
-rwxr-xr-x  1 1001 cnb  3014 Jan  1  1980 python3.11-config

cnb@6d6e72e205b2:/workspace$ /layers/paketo-buildpacks_cpython/cpython/bin/python3.11
/layers/paketo-buildpacks_cpython/cpython/bin/python3.11: error while loading shared libraries: libpython3.11.so.1.0: cannot open shared object file: No such file or directory

cnb@6d6e72e205b2:/workspace$ python3 -V
Python 3.10.12
```

**Comment by hwakabh** [comment]
Since we have been changed builder from paketo to google-cloud in #77 ([commits](https://github.com/hwakabh/bennu-official/pull/77/commits/cbb1988abd7eae0efdebd7cee6aa3c523b346764)), this issue might be staled.
We have also confirmed with `/cnb/process/shell`, which is provided by multiple entries in Procfile, that the CPython version of build artifacts would be same as the one of `.python-version` in code basis.

```shell
(Python:3.12.1 .venv) 24-01-11 13:31:15 git/bennu-official [main] % kubectl -n bennuhp get pods
NAME                      READY   STATUS    RESTARTS   AGE
django-6d486d6857-9985r   1/1     Running   0          10h
mysql-5b59dd6949-2bxkb    1/1     Running   0          45h
(Python:3.12.1 .venv) 24-01-11 13:31:21 git/bennu-official [main] % kubectl -n bennuhp exec -it django-6d486d6857-9985r -- /cnb/process/shell
Defaulted container "django" out of: django, collectstatic (init), migrate (init), loaddata (init)
Python 3.12.1 (main, Dec 13 2023, 05:39:35) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> 
>>> exit()
```
